clientd3d/d3drender.c

```diff
@@ -6909,14 +6879,34 @@ void D3DRenderObjectsDraw(d3d_render_pool_new *pPool, room_type *room,
 		if (pRNode->obj.id == player.id)
 			continue;
 
-		if (flags & OF_INVISIBLE)
+		if (processedIds.find(pRNode->obj.id) == processedIds.end())
+			processedIds.insert(pRNode->obj.id);
+		else
 		{
-			if (GetDrawingEffect(pRNode->obj.flags) != OF_INVISIBLE)
+			objects_skipped++;
+			continue;
+		}
+
+		// Check for invisible objects
+		if((flags & OF_INVISIBLE) == OF_INVISIBLE)
+		{
+			if ((GetDrawingEffect(pRNode->obj.flags) & OF_INVISIBLE) != OF_INVISIBLE)
 				continue;
 		}
 		else
 		{
-			if (GetDrawingEffect(pRNode->obj.flags) == OF_INVISIBLE)
+			if ((GetDrawingEffect(pRNode->obj.flags) & OF_INVISIBLE) == OF_INVISIBLE)
+				continue;
+		}

```

Comment:
Weird logic here.. looks like checking GetDrawingEffect is not needed.

Reply:
Hard to read this.. suggestion:
```
bool objInvisible = (pRNode->obj.flags & OF_INVISIBLE) == OF_INVISIBLE);
```
and use that in the checks.  The logic could further be simplified, I think to an XOR, but up to you whether that's a good idea.